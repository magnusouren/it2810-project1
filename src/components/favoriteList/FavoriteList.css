#favorites-heading {
  margin-bottom: 0;
}
