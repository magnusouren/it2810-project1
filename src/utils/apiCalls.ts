import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
