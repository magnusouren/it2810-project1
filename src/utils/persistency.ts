import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
