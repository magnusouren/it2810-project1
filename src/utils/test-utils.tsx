import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
