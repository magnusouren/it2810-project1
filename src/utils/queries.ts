import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
