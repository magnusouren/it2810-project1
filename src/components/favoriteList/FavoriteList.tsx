import './FavoriteList.css';

import { useQueries } from '@tanstack/react-query';
import { FC } from 'react';

import { Drink } from '../../types';
import { fetchDrinkById } from '../../utils/queries';
import { FavoriteCard } from './favoriteCard/FavoriteCard';

interface FavoriteListProps {
  favorites: string[];
}

export const FavoriteList: FC<FavoriteListProps> = ({ favorites }) => {
  const userQueries = useQueries<Drink[]>({
    queries:
      favorites?.map((favorite) => {
        return {
          queryKey: ['drinkI', favorite], // TODO - felles query for alle drinkQueries i utils
          queryFn: () => fetchDrinkById(favorite).then((res) => res),
        };
      }) || [],
  });

  const data: Drink[] = userQueries.map((query) => query.data as Drink);

  const isLoading: boolean = userQueries.some((query) => query.isLoading);
  const isError: boolean = userQueries.some((query) => query.isError);
  const isSuccess: boolean = userQueries.some((query) => query.isSuccess);

  if (isLoading) return <div>Loading...</div>;
  if (isError) return <div>Something went wrong...</div>;
  if (data.length == 0) return <div>No favorites found...</div>;

  return (
    <>
      <h2 id={'favorites-heading'}>Your favorites:</h2>
      <div className='favorites'>
        <ul>{isSuccess && favorites && data.map((drink, index) => <FavoriteCard drink={drink} key={index} />)}</ul>
      </div>
    </>
  );
};
