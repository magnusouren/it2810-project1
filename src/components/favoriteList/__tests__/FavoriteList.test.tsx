import { screen, waitFor } from '@testing-library/react';
import { vi } from 'vitest';

import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { FavoriteList } from '../FavoriteList';

describe('FavoriteList', () => {
  it('should render correctly when having 2 favorites', async () => {
    const { container } = renderWithRouterAndQueryClient(
      <FavoriteList favorites={['1', '2']} onRemoveFavorite={vi.fn()} />,
    );
    await waitFor(() => screen.getByText('Super Drink'));
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when having 1 favorite', async () => {
    const { container } = renderWithRouterAndQueryClient(<FavoriteList favorites={['1']} onRemoveFavorite={vi.fn()} />);
    await waitFor(() => screen.getByText('Super Drink'));
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when having no favorite', async () => {
    const { container } = renderWithRouterAndQueryClient(<FavoriteList favorites={[]} onRemoveFavorite={vi.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it('should contain heading-element', async () => {
    const { getByRole } = renderWithRouterAndQueryClient(<FavoriteList favorites={['1']} onRemoveFavorite={vi.fn()} />);
    await waitFor(() => screen.getByText('Super Drink'));
    expect(getByRole('heading', { name: 'Your favorites:' })).toBeDefined();
  });

  it('should contain correct amount of favorite-cards', async () => {
    const { getAllByRole } = renderWithRouterAndQueryClient(
      <FavoriteList favorites={['1', '2']} onRemoveFavorite={vi.fn()} />,
    );
    await waitFor(() => screen.getAllByText('Super Drink'));
    expect(getAllByRole('listitem').length).toBe(2);
  });
});
