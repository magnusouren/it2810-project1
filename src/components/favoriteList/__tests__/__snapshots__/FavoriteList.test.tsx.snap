// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteList > should render correctly when having 1 favorite 1`] = `
<div>
  <h2
    id="favorites-heading"
  >
    Your favorites:
  </h2>
  <div
    class="favorites"
  >
    <ul>
      <li>
        <img
          alt="Super Drink"
          src="https://localhost:3000/drink/11118.jpg"
        />
        <div>
          <a
            href="/drink/11118"
          >
            <h3>
              Super Drink
            </h3>
          </a>
          <p
            class="extra-info"
          >
            Category
             - 
            Highball
          </p>
        </div>
        <button
          class="remove-favorite"
        >
          <span
            class="material-symbols-outlined"
          >
            heart_minus
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`FavoriteList > should render correctly when having 2 favorites 1`] = `
<div>
  <h2
    id="favorites-heading"
  >
    Your favorites:
  </h2>
  <div
    class="favorites"
  >
    <ul>
      <li>
        <img
          alt="Super Drink"
          src="https://localhost:3000/drink/11118.jpg"
        />
        <div>
          <a
            href="/drink/11118"
          >
            <h3>
              Super Drink
            </h3>
          </a>
          <p
            class="extra-info"
          >
            Category
             - 
            Highball
          </p>
        </div>
        <button
          class="remove-favorite"
        >
          <span
            class="material-symbols-outlined"
          >
            heart_minus
          </span>
        </button>
      </li>
      <li>
        <img
          alt="Foo bar"
          src="https://localhost:3000/drink/2.jpg"
        />
        <div>
          <a
            href="/drink/22222"
          >
            <h3>
              Foo bar
            </h3>
          </a>
          <p
            class="extra-info"
          >
            Category
             - 
            Foo
          </p>
        </div>
        <button
          class="remove-favorite"
        >
          <span
            class="material-symbols-outlined"
          >
            heart_minus
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`FavoriteList > should render correctly when having no favorite 1`] = `
<div>
  <div>
    No favorites found...
  </div>
</div>
`;
